ddl/table.go

```diff
@@ -1040,11 +1041,32 @@ func onModifyTableCharsetAndCollate(d *ddlCtx, t *meta.Meta, job *model.Job) (ve
 		return ver, errors.Trace(err)
 	}
 
-	tblInfo.Charset = toCharset
-	tblInfo.Collate = toCollate
-
 	if needsOverwriteCols {
+		oldCols := make([]*model.ColumnInfo, 0, len(tblInfo.Columns))
+		newCols := make([]*model.ColumnInfo, 0, len(tblInfo.Columns))
 		// update column charset.
+		for _, col := range tblInfo.Columns {
+			newCol := col.Clone()
+			newCol.SetCharset(toCharset)
+			newCol.SetCollate(toCollate)
+			if field_types.HasCharset(&col.FieldType) && needCheckLatin1Convert(col, newCol) {
+				oldCols = append(oldCols, col)
+				newCols = append(newCols, newCol)
+			}
+		}
+		if len(oldCols) != 0 {
+			err := w.checkLatin1Convert(dbInfo.Name, tblInfo.Name, oldCols, newCols)
+			if err != nil {
+				if job.ReorgMeta.SQLMode.HasStrictMode() {
+					job.State = model.JobStateCancelled
+					return ver, err
+				} else {
+					warn := errors.Cause(err).(*terror.Error)
+					job.ReorgMeta.Warnings[warn.ID()] = warn
+					job.ReorgMeta.WarningsCount[warn.ID()] = 1
+				}
+			}
```

Comment:
We can extract the above code to deal with this problem(add a jobState argument), then we can use this function in https://github.com/pingcap/tidb/pull/34534/files#diff-42e0180cdc8c14def674a9e45d908d6700f6a8ccba0f9ff4b14b27c0daa49859R1640-R1650 too.

Reply:
Done